### ISSUE TITLE
Proposal: speedup screenshots/files transfer

### ISSUE THREAD

**Issue by artem-zinnatullin** [square/spoon#395]
At the moment Spoon grabs screenshots and files after all tests finish running. So if you have a lot of screenshots as we do, it can take minutes in the end of the build to grab them and eventually become comparable with whole build time.

Proposal is to grab screenshots **during** test run, options I see:

* Grab screenshots/files for this exact test when it finishes (push model).
* Grab all screenshots/files periodically (pull model).
* Mixed model.

If you're open to this, let's discuss implementation and I can start working on a PR.

**Comment by edenman** [comment]
I like the push model approach better, I guess?  The pull model seems like it would be potentially racy: the test runner would be writing files and we'd potentially be pulling images before they were finished writing?  That said, I'm not sure how you're going to communicate the "filenames for this particular test" from `Spoon` (which creates the filenames and writes to the files) to `SpoonTestRunner` (which does the pulling)

**Comment by edenman** [comment]
Sharding support is also going to be tricky.

**Comment by artem-zinnatullin** [comment]
>I'm not sure how you're going to communicate the "filenames for this particular test" from `Spoon`

Looks like you meant `Spoon` library which *can* be included as a dependency to the test code while I'm talking about `Spoon` as a test runner which runs on the computer and basically commands which tests to run on which devices and knowns when particular test has finished its execution.

>Sharding support is also going to be tricky.

Well, tricky but seems doable.

**Comment by artem-zinnatullin** [comment]
I also like push model better because it seems more robust to race condition between writing files and grabbing them, thought it can be solved in the `Spoon` library via writing to temp file first.

**Comment by edenman** [comment]
Yeah, sorry, I mean that because the test is the one that determines how many files get written and what the filenames are, the test runner will have to somehow figure out what those filenames are.

tl;dr go for it!

**Comment by otiasj** [comment]
For speeding up the transfer, it would be nice to be able to change the default file format to jpg.
I see in Spoon.takeScreenshot :
bitmap.compress(PNG, 100 /* quality */, fos);
to something like 
bitmap.compress(JPG, 80, fos);

this would reduce the size of files to transfer.

**Comment by tir38** [comment]
@otiasj  I imagine that's a separate feature request. Feel free to create a new github issue.
